csharp/src/Apache.Arrow/Arrays/BinaryArray.cs

```diff
@@ -173,11 +245,19 @@ public TBuilder Set(int index, byte value)
                 throw new NotImplementedException();
             }
 
+            /// <summary>
+            /// Clear all contents appended so far.
+            /// </summary>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
             public TBuilder Clear()
             {
                 ValueOffsets.Clear();
                 ValueBuffer.Clear();
                 ValidityBuffer.Clear();
+
+                // Always write the first offset before anything has been written.
+                Offset = 0;
```

Comment:
Not resetting the `Offset` member variable was the root cause of the string-clearing bug referenced in ARROW-8344.